Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,462 @@
+}
+
+- (CGSize)viewportSize
+{
+    UIInterfaceOrientation orientation = [self orientation];
+    CGSize viewport_size = CGSizeMake(width, height);
+    if (orientation == UIInterfaceOrientationPortrait || orientation == UIInterfaceOrientationPortraitUpsideDown)
+    {
+        if (width > height)
+        {
+            viewport_size = CGSizeMake(height, width);
+        }
+    }
+    else
+    {
+        if (width < height)
+        {
+            viewport_size = CGSizeMake(height, width);
+        }
+    }
+    return viewport_size;
+}
+
+- (void)session:(ARSession *)session didUpdateFrame:(ARFrame *)frame
+{
+    CVPixelBufferRef pixel_buffer = frame.capturedImage;
+    if (CVPixelBufferGetPlaneCount(pixel_buffer) < 2)
+    {
+        return;
+    }
+    
+    _cameraTexture = [self updateCapturedTexture:pixel_buffer];
+    [self updateCamera:frame.camera];
+}
+
+- (void)updateCamera:(ARCamera*)camera
+{
+    auto& frameView = activeFrameViews->at(0);
+    
+    UIInterfaceOrientation orientation = [self orientation];
+    auto transform = [camera transform];
+    auto transformOrientation = simd_quaternion(transform);
+    frameView.Space.Pose.Orientation = {transformOrientation.vector.x
+        , transformOrientation.vector.y
+        , transformOrientation.vector.z
+        , transformOrientation.vector.w};
+    
+    frameView.Space.Pose.Position = { transform.columns[0][3]
+        , transform.columns[1][3]
+        , transform.columns[2][3]};
+    
+    auto projection = [camera projectionMatrixForOrientation:orientation viewportSize:[self viewportSize] zNear:frameView.DepthNearZ zFar:frameView.DepthFarZ];
+    float a = projection.columns[0][0];
+    float b = projection.columns[1][1];
+    float aspectRatio = b/a;
+    float fov = atanf(1.f / b);
+    
+    frameView.FieldOfView.AngleDown = -(frameView.FieldOfView.AngleUp = fov);
+    frameView.FieldOfView.AngleLeft = -(frameView.FieldOfView.AngleRight = fov * aspectRatio);
+    NSLog(@"Update Camera");
+}
+
+@end
+namespace xr
+{
+    class System::Impl
+    {
+    public:
+        Impl(const std::string& applicationName)
+        {
+        }
+
+        bool IsInitialized() const
+        {
+            return true;
+        }
+
+        bool TryInitialize()
+        {
+            return true;
+        }
+    };
+
+    class System::Session::Impl
+    {
+    public:
+        const System::Impl& HmdImpl;
+        std::vector<Frame::View> ActiveFrameViews{ {} };
+        std::vector<Frame::InputSource> InputSources;
+        float DepthNearZ{ DEFAULT_DEPTH_NEAR_Z };
+        float DepthFarZ{ DEFAULT_DEPTH_FAR_Z };
+        bool SessionEnded{ false };
+        id<MTLDevice> MetalDevice;
+        CAMetalLayer* MetalLayer;
+        ARSession* session{};
+        SessionDelegate* sessionDelegate{};
+        ARWorldTrackingConfiguration* configuration{};
+        id<MTLRenderPipelineState> pipelineState;
+        vector_uint2 viewportSize;
+        id<MTLCommandQueue> commandQueue;
```

Comment:
Do all these need to be public? I would guess some/most could be private?

Reply:
moved to private.